core/src/main/java/com/scalar/db/common/ProjectedResult.java

```diff
@@ -0,0 +1,146 @@
+package com.scalar.db.common;
+
+import com.google.common.collect.ImmutableSet;
+import com.scalar.db.api.Result;
+import com.scalar.db.io.Column;
+import com.scalar.db.io.Key;
+import com.scalar.db.io.Value;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** An implementation of {@code Result} that only includes projected columns. */
+@Immutable
+public class ProjectedResult extends AbstractResult {
+  private static final Logger logger = LoggerFactory.getLogger(ProjectedResult.class);
+
+  private final Result original;
+  private final ImmutableSet<String> containedColumnNames;
+
+  public ProjectedResult(Result original, List<String> projections) {
+    this.original = Objects.requireNonNull(original);
+
+    ImmutableSet.Builder<String> builder = ImmutableSet.builder();
+    original.getContainedColumnNames().stream()
+        .filter(c -> projections.isEmpty() || projections.contains(c))
+        .forEach(builder::add);
+    containedColumnNames = builder.build();
+  }
+
+  /** @deprecated As of release 3.8.0. Will be removed in release 5.0.0 */
+  @Deprecated
+  @Override
+  public Optional<Key> getPartitionKey() {
+    return getKey(original.getPartitionKey());
+  }
+
+  /** @deprecated As of release 3.8.0. Will be removed in release 5.0.0 */
+  @Deprecated
+  @Override
+  public Optional<Key> getClusteringKey() {
+    return getKey(original.getClusteringKey());
+  }
+
+  private Optional<Key> getKey(Optional<Key> key) {
+    if (!key.isPresent()) {
+      return Optional.empty();
+    }
+    for (Value<?> value : key.get()) {
+      if (!containedColumnNames.contains(value.getName())) {
+        logger.warn("Full key doesn't seem to be projected into the result");
```

Comment:
In this case, it means `Cassandra.scan()` or equivalents didn't setup proper projections? If so, it's not an expected situation and throwing an AssertionError or IllegalStateException would be good, I think.

Reply:
Honestly, I just followed the same manner as `FilteredResult` in the consensus commit, but `IllegalStateException` with a message like `CoreError.COLUMN_NOT_FOUND` would be better as you mentioned.

@brfrn169 Do you remember why you chose a warning rather than an exception?

Reply:
@komamitsu Thank you for the suggestion. I fixed it, including other similar classes in this PR based on the discussion with @brfrn169.